parsers/UY.py

```diff
@@ -1,163 +1,224 @@
+    next_day_format = "{day}%2F{month}%2F{year}".format(
+        day=next_day.day, month=next_day.month, year=next_day.year
     )
 
-    obj = {"datetime": date.datetime}
-
-    for tr in table.find_all("tr"):
-        tds = tr.find_all("td")
-        if not len(tds):
-            continue
-
-        key = tds[0].find_all("b")
-        # Go back one level up if the b tag is not there
-        if not len(key):
-            key = tds[0].find_all("font")
-        k = key[0].contents[0]
-
-        value = tds[1].find_all("b")
-        # Go back one level up if the b tag is not there
-        if not len(value):
-            value = tds[1].find_all("font")
-        v_str = value[0].contents[0]
-        if v_str.find(",") > -1 and v_str.find(".") > -1:
-            # there can be values like "1.012,5"
-            v_str = v_str.replace(".", "")
-            v_str = v_str.replace(",", ".")
-        else:
-            # just replace decimal separator, like "125,2"
-            v_str = v_str.replace(",", ".")
-        v = float(v_str)
-
-        # solar reports -0.1 at night, make it at least 0
-        v = max(v, 0)
-
-        obj[k] = v
-
-    # https://github.com/tmrowco/electricitymap/issues/1325#issuecomment-380453296
-    salto_grande = get_salto_grande(session)
-    obj["Hidráulica"] = obj.get("Hidráulica", 0.0) + salto_grande
-
-    return obj
-
-
+    link = ADME_URL + date_format + "&fecha_fin=" + next_day_format + "&send=MOSTRAR"
+    r = session.get(url=link)
+    soup = BeautifulSoup(r.content, "html.parser")
+    href_tags = soup.find_all("a", href=True)
+    for tag in href_tags:
+        if tag.button is not None:
+            if tag.button.string == "Archivo Scada Detalle 10minutal":
+                data_url = "https://pronos.adme.com.uy" + tag.get("href")
+    return data_url
+
+
+def fetch_data(
+    zone_key: str,
+    session: Session,
+    target_datetime: datetime,
+    sheet_name: str,
+    logger: Logger = getLogger(__name__),
+) -> pd.DataFrame:
+    """ """
+    assert target_datetime is not None
+    assert session is not None
+    assert sheet_name != "" or sheet_name is not None
+    adme_url = get_adme_url(target_datetime=target_datetime, session=session)
+    r = session.get(url=adme_url)
+    if r.status_code == 200:
+        df_data = pd.read_excel(
+            io.BytesIO(r.content), engine="odf", header=2, sheet_name=sheet_name
+        )
+        df_data.columns = df_data.columns.str.strip()
+        df_data = df_data.rename(columns={"Fecha": "datetime"})
+
+        df_data = df_data.set_index("datetime")
+        return df_data
+    else:
+        raise ParserException(
+            parser="UY.py",
+            message="no data available for target_dateitme",
+        )
+
+
+def fix_solar_production(dt: datetime, row: pd.Series) -> int:
+    """sets solar production to 0 during the night as there is only solar PV in UY"""
+    if (5 >= dt.hour or dt.hour >= 20) and row.get("value") != 0:
```

Comment:
Isn't that a bit arbitrary ? 😅

Reply:
https://sunrise.maplogs.com/uruguay.120.html?year=2022
I had a look at that website for 2022 and thought it was an ok-ish range.
For some reason there is some solar production at night and it's really weird
![image](https://user-images.githubusercontent.com/107848894/210526364-ec6e62f9-697e-40b0-9ed7-de6549c91db1.png)

Reply:
It's possible this also includes storage from solar generation sites such as battery storage but that's just a guess.

Reply:
I would argue that we should set to 0 if we don't know ( the quantities are small anyways)

Reply:
Could it be a timezone issue?

Reply:
Solar could also include Concentrated Solar Power with energy storage. I found [this 2015 government study](http://www.energiasolar.gub.uy/index.php/investigacion-e-innovacion/investigacion-concentradores) investigating it, but haven't encountered anything they actually went through with it.